pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,1301 @@
+              num_local_qubits),
+          numGlobalQubits_(num_global_qubits),
+          numLocalQubits_(num_local_qubits), mpi_manager_(mpi_communicator),
+          handle_(make_shared_cusv_handle()),
+          cublascaller_(make_shared_cublas_caller()),
+          localStream_(make_shared_local_stream()),
+          svSegSwapWorker_(make_shared_mpi_worker<CFP_t>(
+              handle_.get(), mpi_manager_, BaseType::getData(),
+              num_local_qubits, localStream_.get())),
+          gate_cache_(true) {
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize())
+        mpi_manager_.Barrier();
+    };
+
+    StateVectorCudaMPI(size_t num_global_qubits, size_t num_local_qubits)
+        : StateVectorCudaBase<Precision, StateVectorCudaMPI<Precision>>(
+              num_local_qubits),
+          numGlobalQubits_(num_global_qubits),
+          numLocalQubits_(num_local_qubits), mpi_manager_(MPI_COMM_WORLD),
+          handle_(make_shared_cusv_handle()),
+          cublascaller_(make_shared_cublas_caller()),
+          localStream_(make_shared_local_stream()),
+          svSegSwapWorker_(make_shared_mpi_worker<CFP_t>(
+              handle_.get(), mpi_manager_, BaseType::getData(),
+              num_local_qubits, localStream_.get())),
+          gate_cache_(true) {
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize())
+        mpi_manager_.Barrier();
+    };
+
+    ~StateVectorCudaMPI(){};
+
+    /**
+     * @brief Get the total number of wires.
+     */
+    auto getTotalNumQubits() const {
+        return numGlobalQubits_ + numLocalQubits_;
+    }
+
+    /**
+     * @brief Get the number of wires distributed across devices.
+     */
+    auto getNumGlobalQubits() const { return numGlobalQubits_; }
+
+    /**
+     * @brief Get the number of wires within the local devices.
+     */
+    auto getNumLocalQubits() const { return numLocalQubits_; }
+
+    /**
+     * @brief Get pointer to custatevecSVSwapWorkerDescriptor.
+     */
+    auto getSwapWorker() -> custatevecSVSwapWorkerDescriptor_t {
+        return svSegSwapWorker_.get();
+    }
+    /**
+     * @brief Init 00....0>.
+     */
+    void initSV_MPI(bool async = false) {
+        size_t index = 0;
+        const std::complex<Precision> value = {1, 0};
+        BaseType::getDataBuffer().zeroInit();
+        setBasisState(value, index, async);
+    }
+
+    /**
+     * @brief Set value for a single element of the state-vector on device. This
+     * method is implemented by cudaMemcpy.
+     *
+     * @param value Value to be set for the target element.
+     * @param index Index of the target element.
+     * @param async Use an asynchronous memory copy.
+     */
+    void setBasisState(const std::complex<Precision> &value, const size_t index,
+                       const bool async = false) {
+        size_t rankId = index >> BaseType::getNumQubits();
+
+        size_t local_index = (rankId << BaseType::getNumQubits()) ^ index;
```

Comment:
Theoretically yes, but practically no since the maximum number of qubits we can reach now is around 40. To make it safer, updated with `std::pow` here instead!